### ISSUE TITLE
[HOLD for payment 2023-06-19] [$1000] Checkbox border turns green once another window is opened and revert back

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#18801]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open the Settings menu.
2. Click on " Workspaces."
3. Open any Workspace.
4. Select "Manage Members."
5. Click "Select all."
6. In the presence of admin only as a member observe the checkbox, green border appears for a fraction of second, when clicked.
7. Open another window and observe the checkbox border turns green permanently till clicked again.

## Expected Result: 
The checkbox should always have a white border when the admin is the only member present, as the admin cannot be selected when using the "Select all" option.
## Actual Result: 
The border of checkbox "Select all" turns green once another window is opened and reverted back.
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.13.0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/40d25b9d-9193-42b2-bc40-3c06f0992a42

https://github.com/Expensify/App/assets/43996225/5f2c386f-5561-41ea-aead-638d5c8d1302

**Expensify/Expensify Issue URL:**
**Issue reported by:** @usmantariq96
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1683739504158529

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01f763eb9e34751150</li>
        <li>Upwork Job ID: 1659525973736022016</li>
        <li>Last Price Increase: 2023-06-02</li>
    </ul>
</details>

**Comment by hungvu193** [comment]
This is decided is not a bug here: https://github.com/Expensify/App/issues/11283#issuecomment-1292300799
cc @parasharrajat

**Comment by mallenexpensify** [comment]
Oooooof, I feel like there are so many small, edge case, possible-bugs that require a bunch of random-ish steps in order to reproduce.  Like... if we didn't fix this, would any user/customer ever think it was a bug? (end.. rant). 

I'll come back to this tomorrow

**Comment by mallenexpensify** [comment]
> I think adding a Low priority label would help Melvin to understand the situation maybe?

Agreed but... then we'd like just end up with a lot of these lil bugs never getting fixed then we'd have an app with ton of tiny bugs!  (well.. maybe.  I'm way less _grumpy_ today, glad I slept on it). 

So.. is this a small and edge case bug, yup. 
Should we fix it... yes, unless there's a technical reason we might not want to that I'm unaware of.  👀 plz @MariaHCD to see if you agree, if so, please add `External` if you think it can be worked on by a contributor.

**Comment by MariaHCD** [comment]
I definitely agree that this is a small, edge case situation. I don't think a customer would actually notice but it is odd that the checkbox would still be highlighted when the user navigates away.

So, I agree with @allroundexperts, if the element wasn't highlighted before the tab switch, then it shouldn't be highlighted after.

While I don't necessarily think this is a bug since it's more of an small improvement, I still think we should fix it. Like @mallenexpensify said we don't want to end up with an app that has tiny, weird issues like these.

**Comment by leotaylorfan** [comment]
Hi, I am interested in fixing that bug.
How can I access the project?
Let me know the way.

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
A previously unchecked checkbox will show it focus outline after switching tabs.

### What is the root cause of that problem?
When we switch the tab back to the app, it will refocus the previously focused element, in this case the checkbox. Even though the checkbox lose it's outline when unchecked, the checkbox is still in focus state, which I think could lead to confusion. When the user uncheck the checkbox, we don't want it to lose it's focus because it will mess the keyboard navigation.

If we look at the initial [PR](https://github.com/Expensify/App/pull/8874) implementation of the checkbox, the PR is trying to achieve accessibility with keyboard for the checkbox but the checkbox missing the outline when navigated with tab, thus the PR added the outline (alongside other changes) when the checkbox received focus. This later create the solved issue we have [here](https://github.com/Expensify/App/issues/11283) and now this one.

### What changes do you think we should make in order to solve the problem?

1. As the initial intention is for the accessibility, we can remove the current implementation of the `isFocused` state in checkbox. 
2. Then, we need to show the outline only if the checkbox receive focus by keyboard, which can be achieve with `:focus-visible`
Currently, rn-web [does not support](https://www.github.com/necolas/react-native-web/issues/1849) it yet and I'm not sure if there is a reliable way to detect whether the focus event comes from mouse or keyboard. So, we need to do it natively by adding the :focus-visible style only to checkbox (in `index.html`) Here is how it looks:

```
div[role="checkbox"]:focus-visible > div {
    border-color: #03D47C !important;
}
```

`> div` so the inner view receive the border style (`styles.borderColorFocus`).
https://github.com/Expensify/App/blob/8a3d79cde0255b2abb7a8f4f518ec22834a54203/src/components/Checkbox.js#L116-L134

Otherwise, do nothing I guess.

**Comment by mallenexpensify** [comment]
Thanks @parasharrajat .  Is this an example where we should 
1. Update the design guidelines doc for checkboxes
2. Review all closed issues regarding checkboxes to ensure they're aligned with the guidelines
3. Review all open issues, point to the guidelines, close/merge issues if needed
4. Re-review this issue. 

I feel like we have a gap for design-related instances like this, where we don't have guidelines or a _source of truth_ to point to.

**Comment by mananjadhav** [comment]
@mallenexpensify @MariaHCD I don't think I can get to this soon. Can you please assign another C+ here?

**Comment by s77rt** [comment]
taking this!

**Comment by s77rt** [comment]
We can also hold this issue for https://github.com/necolas/react-native-web/issues/1849 but this feature reuqest is pretty old and not being actively worked on. But it would be really great if we have it, all we would have to do is pass `styles.borderColorFocus` if `focusVisible` is true (we would still remove the isFocused state management).

**Comment by bernhardoj** [comment]
> We want the border color to be applied only if we are using [the standard checkbox view](https://github.com/Expensify/App/blob/b02d90291e277bdcb93961048a25a31fb8060198/src/components/Checkbox.js#L116-L134).

@s77rt You're right. Honestly it's weird to me a checkbox can receive a custom view 😄. I see we use it for the password eye icon that basically just being used as a normal pressable. Anyway, I was looking for a way for the CheckBox VIew to have some "id" so we can select it in CSS. Using ID sounds invalid by the spec.

> We can also hold this issue for https://github.com/necolas/react-native-web/issues/1849

I saw that too and looks like they don't have any plan for that now.

**Comment by s77rt** [comment]
@bernhardoj I don't think we can use ID as we may have two checkboxes in same page.

**Comment by bernhardoj** [comment]
Updated by using `data-*` as the selector.

**Comment by s77rt** [comment]
@bernhardoj Thanks! Still not the cleanest but the base approach is better and correct. That looks good enough to me.
@MariaHCD What are your thoughts here?

**Comment by s77rt** [comment]
I think we can fix this in another way that may look cleaner but the outline will be blue as the rest of the elements and not green.

**Comment by s77rt** [comment]
Not overdue. @MariaHCD ^

**Comment by s77rt** [comment]
ah I'm not the C+ yet :sweat_smile:

**Comment by MariaHCD** [comment]
Whoops, just meant to assign the C+

**Comment by MariaHCD** [comment]
> @MariaHCD What are your thoughts here?

Yeah, I like @bernhardoj's RCA and although having the css in index.html isn't super ideal, I think the solution makes sense. 

@s77rt Do you think we should wait for more proposals here?

**Comment by s77rt** [comment]
@MariaHCD What we are looking here is to adjust the focus-visible style. This can't be done (in a clean way) using React itself - it is not supported. We have to do it externally. Actually we already do this but in a generic way. https://github.com/Expensify/App/blob/967100432a5f3ae5458f43ab20163ea07f6d17dc/web/index.html#L55-L58

The generic one is not working on the checkbox due to it's structure - the blue outline is under the box. We can fix that but the outline will be blue and not green.

Maybe we should fix the missing blue outline first? Then figure out a way to change the color?

**Comment by MariaHCD** [comment]
@s77rt couple clarifications here :)

> The generic one is not working on the checkbox due to it's structure

Could you clarify what you mean by "structure"?

> Maybe we should fix the missing blue outline first?

So you mean:
1. Remove the current green outline when the checkbox is focused
2. Make sure the blue outline works for checkboxes
3. Find a way to change that outline to green only for checkboxes

Do I have that right? If so, 3 would involve some custom css as well, right? So would it be a cleaner approach?

**Comment by s77rt** [comment]
> Could you clarify what you mean by "structure"?

The box shadow is inset. It's inside the div. But in the div the checkbox is taking the whole space thus the shadow is not visible.

| Inset | Outset | Inset (with padding) |
|-------|--------|----------------------|
|  <img src="https://github.com/Expensify/App/assets/16493223/00ee1858-7d3d-4c55-ab81-c8ab50314900" />     |     <img src="https://github.com/Expensify/App/assets/16493223/921cae72-c9ad-4da4-8e81-4e3d4bf4bbeb" />   |         <img src="https://github.com/Expensify/App/assets/16493223/45adec78-62fe-4c21-bcc2-ac8139c2f853" />             |


___

> Do I have that right? If so, 3 would involve some custom css as well, right? So would it be a cleaner approach?

Yes, but the approach won't be much cleaner. We will still need some extra filter condition to apply the green outline on checkbox only.

**Comment by MariaHCD** [comment]
Nice, thanks for clarifying!

> Yes, but the approach won't be much cleaner. We will still need some extra filter condition to apply the green outline on checkbox only.

Understood, in this case, @bernhardoj's solution seems simpler but maybe if we see the code changes that we would need for this approach, we can make a better decision. What do you think?

**Comment by s77rt** [comment]
@MariaHCD The code changes is basically what @bernhardoj mentioned in addition to the removal of the `isFocused` state and `onBlur`/`onFocus` callbacks.


<details>
<summary>Code diff</summary>

```diff
diff --git a/src/components/Checkbox.js b/src/components/Checkbox.js
index 9e4f5da57d..a9cd24b044 100644
--- a/src/components/Checkbox.js
+++ b/src/components/Checkbox.js
@@ -50,24 +50,11 @@ const defaultProps = {
 class Checkbox extends React.Component {
     constructor(props) {
         super(props);
-        this.state = {
-            isFocused: false,
-        };
 
-        this.onFocus = this.onFocus.bind(this);
-        this.onBlur = this.onBlur.bind(this);
         this.handleSpaceKey = this.handleSpaceKey.bind(this);
         this.firePressHandlerOnClick = this.firePressHandlerOnClick.bind(this);
     }
 
-    onFocus() {
-        this.setState({isFocused: true});
-    }
-
-    onBlur() {
-        this.setState({isFocused: false});
-    }
-
     handleSpaceKey(event) {
         if (event.code !== 'Space') {
             return;
@@ -83,13 +70,6 @@ class Checkbox extends React.Component {
             return;
         }
 
-        const wasChecked = this.props.isChecked;
-
-        // If checkbox is checked and focused, make sure it's unfocused when pressed.
-        if (this.state.isFocused && wasChecked) {
-            this.onBlur();
-        }
-
         this.props.onPress();
     }
 
@@ -99,10 +79,7 @@ class Checkbox extends React.Component {
                 disabled={this.props.disabled}
                 onPress={this.firePressHandlerOnClick}
                 onMouseDown={this.props.onMouseDown}
-                onFocus={this.onFocus}
-                onBlur={this.onBlur}
                 ref={this.props.forwardedRef}
-                onPressOut={this.onBlur}
                 style={this.props.style}
                 onKeyDown={this.handleSpaceKey}
                 accessibilityRole="checkbox"
@@ -120,8 +97,10 @@ class Checkbox extends React.Component {
                             this.props.isChecked && styles.checkedContainer,
                             this.props.hasError && styles.borderColorDanger,
                             this.props.disabled && styles.cursorDisabled,
-                            (this.state.isFocused || this.props.isChecked) && styles.borderColorFocus,
                         ]}
+
+                        // Used as CSS selector to customize focus-visible style
+                        dataSet={{checkbox: true}}
                     >
                         {this.props.isChecked && (
                             <Icon
diff --git a/web/index.html b/web/index.html
index 90011c87e7..ff4b4717f8 100644
--- a/web/index.html
+++ b/web/index.html
@@ -64,6 +64,10 @@
         select:focus-visible, select:focus[data-focusvisible-polyfill]  {
             box-shadow: none;
         }
+        div[role="checkbox"]:focus-visible > div[data-checkbox], div[role="checkbox"]:focus[data-focusvisible-polyfill] > div[data-checkbox] {
+            border-color: #03D47C !important;
+        }
+
         ::-ms-reveal {
             display: none;
         }
```


</details>

**Comment by bernhardoj** [comment]
I'm not sure we can use outset mode as some component outline will not be visible. Example:
<img width="961" alt="image" src="https://github.com/Expensify/App/assets/50919443/3f9bff93-2e0a-45af-a9c4-a592c89ee8cd">
<img width="968" alt="image" src="https://github.com/Expensify/App/assets/50919443/a81d5ee7-625d-4524-ac62-e9d2db3e8a58">
<img width="387" alt="image" src="https://github.com/Expensify/App/assets/50919443/92931cf0-b824-4e60-90da-6c3d953f646c">

**Comment by s77rt** [comment]
@bernhardoj I'm not suggesting the use of an outset shadow, just wanted to clarify why it's not working.

**Comment by s77rt** [comment]
@MariaHCD Correct!

**Comment by MariaHCD** [comment]
Sounds good! How do you want to approach the PR for this, @s77rt? 

Perhaps @bernhardoj, you can incorporate those changes into your proposal (if you agree) and we can assign you to this issue?

**Comment by mallenexpensify** [comment]
> you can incorporate those changes into your proposal (if you agree)

@bernhardoj , can you do so, get _sign off_ from @MariaHCD then we'll hire and assign you?

**Comment by bernhardoj** [comment]
Yes, I can do that.

**Comment by s77rt** [comment]
Melvin we have a proposal already.

**Comment by MariaHCD** [comment]
Friendly bump @bernhardoj :) Would you be able to update your proposal?

**Comment by bernhardoj** [comment]
@MariaHCD sorry, I don't know we were waiting on me. I don't understand what should be updated from my proposal? The code changes from @s77rt is basically based on my proposal already.

**Comment by MariaHCD** [comment]
No worries, @bernhardoj! We just want to see the full proposal here that includes @s77rt's changes. I'd rather have the plan here first vs working it out later in the PR. Does that make sense?

**Comment by bernhardoj** [comment]
Ah, do you mean to include the code into my proposal?

**Comment by MariaHCD** [comment]
Yes :)

**Comment by bernhardoj** [comment]
Added the code. Sorry for the delay.

**Comment by MariaHCD** [comment]
Nice, looks good to me! Thanks @bernhardoj

**Comment by bernhardoj** [comment]
PR is ready.

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified: https://github.com/Expensify/App/pull/13181
- The offending PR has been commented on: https://github.com/Expensify/App/pull/13181#discussion_r1228566410
- A discussion in #expensify-bugs has been started: https://expensify.slack.com/archives/C049HHMV9SM/p1686682753141099
- Determine if we should create a regression test for this bug: I'm leaning to a no, this is pretty minor bug

**Comment by MariaHCD** [comment]
Nice catch, @parasharrajat! @s77rt @bernhardoj let's find a way to have this controlled by our theme variables instead of hardcoding it and create a follow up PR. 

> Do you want to bring it to slack to get a consensus?

I think the consensus on Slack will still be that we shouldn't be hardcoding this. So I think we should first try to find an alternative solution and if we can't, we can move the conversation to Slack.

Thoughts?

**Comment by bernhardoj** [comment]
Here is the alternative solution I find from OpenAI + SO:

```
useEffect(() => {
    const styleElement = document.createElement('style');
    const css = `
    div[role="checkbox"]:focus-visible > div[data-checkbox], div[role="checkbox"]:focus[data-focusvisible-polyfill] > div[data-checkbox] {
        border-color: ${themeColors.borderFocus} !important;
    }
    `;
    styleElement.appendChild(document.createTextNode(css));
    document.head.appendChild(styleElement);
}, []);
```
We will create a new style element and append it to `head` when `App.js` mounts. We can also use this solution for the splash screen background color
https://github.com/Expensify/App/blob/862955c2f15bce3c68d87c549b7af0e27a0eb526/web/index.html#L109-L115

**Comment by s77rt** [comment]
I prefer to keep this hard-coded for now. We already have the blue outline and the splash background hard-coded. Those need to be fixed too and I think such change can be handled once we add the light/dark theme switch feature. Otherwise if you have a strong preference on fixing this now we can probably use `HtmlWebpackPlugin`'s `templateParameters` to pass the theme colours.

**Comment by MariaHCD** [comment]
> I think such change can be handled once we add the light/dark theme switch feature

That's a fair point. Removing those hardcoded colors from index.html might already be a consideration for the `Theme Switching / Light Mode` project.

@shawnborton @grgia Can you confirm if removing hardcoded color values such as for the background color of the splash screen from index.html is something that is or should be part of `Theme Switching / Light Mode`? 

https://github.com/Expensify/App/blob/862955c2f15bce3c68d87c549b7af0e27a0eb526/web/index.html#L109-L115

We introduced another hardcoded color [here](https://github.com/Expensify/App/pull/20252) for focused checkboxes and we'd like to know whether we should fix that now but also don't want to cause any conflicts if there is already a plan in place to replace those hardcoded values.

**Comment by grgia** [comment]
for the hardcoded color on the splash, I believe that will remain theme-independent because it is created before the app is loaded. However, we should avoid hard-coding any colors that are seen after the app is loaded, especially ones that would need to be affected by a theme change. Does that help @MariaHCD?

**Comment by MariaHCD** [comment]
Thanks, @grgia! 

So I think we'll need to fix the hardcoded color now. Thoughts, @bernhardoj @s77rt?

**Comment by s77rt** [comment]
@MariaHCD At this point looks like we have nothing to do. We have this PR https://github.com/Expensify/App/pull/20997 that applied this [discussed solution](https://github.com/Expensify/App/issues/18801#issuecomment-1566955503) i.e [add padding](https://github.com/Expensify/App/issues/18801#issuecomment-1566981603). Except that they didn't change the border color. @shawnborton seems okay with that so looks we all set for now. (The green hardcoded is also removed since it's no longer used and the generic blue one is used instead - the blue color is still hard coded though)

**Comment by mallenexpensify** [comment]
I see a lot of conversation after the PR hit production, any reason I **shouldn't** pay everyone now?  It's been over 7 days

**Comment by s77rt** [comment]
I think it's ready for payment. No further action is required from our end. cc @MariaHCD can you please confirm the same

**Comment by mallenexpensify** [comment]
eeeeps, the old job closed.  I was going to try to get this paid before the weekend. 
@usmantariq96, @bernhardoj , @s77rt , can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~0161807157644d1613

**Comment by s77rt** [comment]
@mallenexpensify Accepted!

**Comment by usmantariq96** [comment]
@mallenexpensify offer accepted

**Comment by bernhardoj** [comment]
@mallenexpensify accepted

**Comment by MariaHCD** [comment]
> We have this PR #20997 that applied this [discussed solution](https://github.com/Expensify/App/issues/18801#issuecomment-1566955503) i.e [add padding](https://github.com/Expensify/App/issues/18801#issuecomment-1566981603).

Perfect, thanks! Yes, looks like we're good to close this out.

**Comment by mallenexpensify** [comment]
Thanks all, everyone is paid, I added this to [our design guideline sheet](https://docs.google.com/spreadsheets/d/1VTQKn9t9xNMbrhD46OIf3fJGmpH4Gmi42jXjQgmtBoY/edit#gid=1045807246&range=A5)
